exchange/accounts/accounts.go

```diff
@@ -0,0 +1,303 @@
+	}
+	return currs, nil
+}
+
+// SubAccounts returns the public SubAccounts and their balances
+// If creds is nil, all credential SubAccounts will be returned
+// If assetType is asset.All, all assets will be returned
+func (a *Accounts) SubAccounts(creds *Credentials, assetType asset.Item) (SubAccounts, error) {
+	if err := common.NilGuard(a); err != nil {
+		return nil, err
+	}
+
+	if !assetType.IsValid() && assetType != asset.All {
+		return nil, fmt.Errorf("%s %s %w", a.Exchange.GetName(), assetType, asset.ErrNotSupported)
+	}
+
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	var subAccts SubAccounts
+	for credsKey, subAccountsForCreds := range a.subAccounts {
+		if !creds.IsEmpty() && *creds != credsKey {
+			continue
+		}
+		for subAcctKey, balances := range subAccountsForCreds {
+			if assetType != asset.All && assetType != subAcctKey.Asset {
+				continue
+			}
+			subAccts = append(subAccts, &SubAccount{
+				ID:        subAcctKey.SubAccount,
+				AssetType: subAcctKey.Asset,
+				Balances:  balances.Public(),
+			})
+		}
+	}
+
+	if len(subAccts) == 0 {
+		return nil, fmt.Errorf("%w for %s credentials %s asset %s", ErrNoSubAccounts, a.Exchange.GetName(), creds, assetType)
+	}
+	return subAccts, nil
+}
+
+// GetBalance returns a copy of the balance for that asset item
+func (a *Accounts) GetBalance(subAccount string, creds *Credentials, aType asset.Item, c currency.Code) (Balance, error) {
+	if err := common.NilGuard(a); err != nil {
+		return Balance{}, err
+	}
+	if !aType.IsValid() {
+		return Balance{}, fmt.Errorf("cannot get balance: %w: %q", asset.ErrNotSupported, aType)
+	}
+
+	if creds.IsEmpty() {
+		return Balance{}, fmt.Errorf("cannot get balance: %w", errCredentialsEmpty)
+	}
+
+	if c.IsEmpty() {
+		return Balance{}, fmt.Errorf("cannot get balance: %w", currency.ErrCurrencyCodeEmpty)
+	}
+
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	subAccts, ok := a.subAccounts[*creds]
+	if !ok {
+		return Balance{}, fmt.Errorf("%w for %s", ErrNoBalances, creds)
+	}
+
+	assets, ok := subAccts[key.SubAccountAsset{
+		SubAccount: subAccount,
+		Asset:      aType,
+	}]
+	if !ok {
+		return Balance{}, fmt.Errorf("%w for %s SubAccount %q %s", ErrNoBalances, a.Exchange.GetName(), subAccount, aType)
+	}
+	b, ok := assets[c.Item]
+	if !ok {
+		return Balance{}, fmt.Errorf("%w for %s SubAccount %q %s %s", ErrNoBalances, a.Exchange.GetName(), subAccount, aType, c)
+	}
+	return b.Balance(), nil
+}
+
+// Save updates the account balances
+// If isSnapshot is true any missing currencies will be removed
+// Credentials will be retrieved from ctx, Use DeployCredentialsToContext.
+// Changes to balances are published individually
+func (a *Accounts) Save(ctx context.Context, subAccts SubAccounts, isSnapshot bool) error {
+	if err := common.NilGuard(a); err != nil {
+		return fmt.Errorf("cannot save holdings: %w", err)
+	}
+	if err := common.NilGuard(a.subAccounts); err != nil {
+		return fmt.Errorf("cannot save holdings: %w", err)
+	}
+
```

Comment:
```suggestion
	if err := common.NilGuard(a, a.subAccounts); err != nil {
		return fmt.Errorf("cannot save holdings: %w", err)
	}

```

Reply:
This is not possible. It would panic on `a` being nil.
I tried in vein to solve this problem and never game up with a solution that wasn't insane.